src/Week6/Fibonacci.hs

```diff
@@ -0,0 +1,310 @@
+ The reason they’re so famous probably has something to do with the
+ simplicity of their definition combined with the astounding variety of
+ ways that they show up in various areas of mathematics as well as art
+ and nature.
+-}
+
+{-
+ Exercise 1:
+
+ Translate the above definition of Fibonacci numbers directly into a recursive
+ function definition of type
+      fib :: Integer -> Integer
+ so that fib n computes the nth Fibonacci number Fn. Now use fib to define
+ the infinite list of all Fibonacci numbers,
+      fibs1 :: [Integer]
+
+ (Hint: You can write the list of all positive integers as [0..].)
+-}
+fib :: Integer -> Integer
+fib 0 = 0
+fib 1 = 1
+fib n = fib (n - 1) + fib (n - 2)
+
+
+fibs1 :: [Integer]
+fibs1 = map fib [0..]
+
+
+-- Another way to write this using list comprehension
+fibs1' :: [Integer]
+fibs1' = [fib i | i <- [0..]]
+
+
+{-
+ Exercise 2:
+
+ When I said “we” in the previous sentence I actually meant “you”.
+ Your task for this exercise is to come up with more efficient implementation.
+ Specifically, define the infinite list
+     fibs2 :: [Integer]
+
+ so that it has the same elements as fibs1, but computing the first n
+ elements of fibs2 requires only O(n) addition operations. Be sure to
+ use standard recursion pattern(s) from the Prelude as appropriate.
+-}
+fibs2 :: [Integer]
+fibs2 = 0 : 1 : fib2 0 1
+  where
+    fib2 a b = (a + b) : fib2 b (a + b)
+
+
+-- We can also rewrite it using list comprehension
+fibs2' :: [Integer]
+fibs2' = 0 : 1 : [fibs2' !! (i - 2) + fibs2' !! (i - 1) | i <- [2..]]
+
+
+{-
+ PREFACE:
+
+ Streams:
+ We can be more explicit about infinite lists by defining a type Stream
+ representing lists that must be infinite. (The usual list type represents
+ lists that may be infinite but may also have some finite length.)
+ In particular, streams are like lists but with only a “cons” constructor—
+ whereas the list type has two constructors, [] (the empty list) and
+ (:) (cons), there is no such thing as an empty stream. So a stream is
+ simply defined as an element followed by a stream.
+-}
+
+{-
+ Exercise 3:
+
+ • Define a data type of polymorphic streams, Stream.
+
+ • Write a function to convert a Stream to an infinite list,
+     streamToList :: Stream a -> [a]
+
+ • To test your Stream functions in the succeeding exercises, it will be
+ useful to have an instance of Show for Streams. However, if you put deriving
+ Show after your definition of Stream, as one usually does, the resulting
+ instance will try to print an entire Stream—which, of course, will never
+ finish. Instead, you should make your own instance of Show for Stream,
+
+    instance Show a => Show (Stream a) where show ...
+
+ which works by showing only some prefix of a stream (say, the first 20 elements).
+-}
+data Stream a = Cons a (Stream a)
+
+
+streamToList :: Stream a -> [a]
+streamToList (Cons x xs) = x : streamToList xs
+
+
+streamFromList :: [a] -> Stream a
+streamFromList xs =
```

Comment:
What happens if `xs` is empty?

Reply:
You're right. This is actually a non-exhaustive pattern as of now. But generating a stream from an empty list didn't really make sense. It is generally a `cons` followed by another stream. Whereas in a list we have `cons` and `[]` (an empty list) constructor.
So for this particular case, it would crash at runtime if we tried generating a stream from an empty list.
Is there a more appropriate way to handle such cases?

Reply:
The issue is, every `Stream a` value has a `[a]` representation, but every `[a]` need not necessarily have a `Stream` representation (since `[a]` values may also be finite). `streamForList` would thus be a partial function, since it is defined over the domain of _all_ `[a]`s. There is thus no obvious or natural way of mapping a `[a]` to ` Stream a`, which is why it isn't a question in this exercise. 

Of course, we can always map lists to infinite lists (using `cycle` as you have done, for example), and then map these lists to their corresponding `Stream` values. But then you have to make sure that you handle the case of the empty list. So I would ask you to consider:

- Is it at all possible to map a _polymorphic_ empty list to an infinite list? If so, then we are done.
- If not, how else might you define `streamFromList` such that it does not potentially crash at runtime?

Reply:
So the initial idea for the function was just to play around list and streams a bit.
But I think this will turnout to be a valuable lesson on how NOT to design functions.

As for the points, I don't think building a polymorphic list is the right way (or a feasible way to speak). Since from what I've read till now, it seems Haskell is intentionally designed to not deal with such lists.
To tackle such problem, I'd redefine `Stream` itself. It could be something like:
```haskell
data Stream = Nil | Cons a (Stream a)

streamToList :: Stream a -> [a]
streamToList Nil = []
streamToList (Cons x xs) = x : streamToList xs


streamFromList :: [a] -> Stream a
streamFromList []     = Nil
streamFromList (x:xs) = Cons x $ streamFromList xs
```

But then it will become essentially a list construct. (Since list has an `Empty` and `cons` constructor)

Also, `Data.Stream` has a similar function https://hackage.haskell.org/package/Stream-0.4.7.2/docs/Data-Stream.html#t:Stream
which essentially fails at an empty list. So, I think there's no _proper_ way to make it not crash at runtime.

Reply:
@nitinprakash96, correct, it will become a type isomorphic to lists, but also defeat the whole point of defining a new type that guarantees that the list is infinite. 

Building a polymorphic list is indeed not the right way to go, but why? More specifically, for the empty list case, we will still need to provide some value of type `a` to construct the stream. Why would this not be possible? 

As for there not being a proper way, you could always use a `Maybe` or `Either`. This would in fact be the preferred approach, and in CDMP, for example, we mostly use total functions (like `safeHead` or `utf8Decode'`) in order to handle the exception more elegantly.

Reply:
So I looked into `Either` and the following came into my mind.
```haskell
streamFromList :: [a] -> Either [a] (Stream a)
streamFromList xs
  | null xs   = Left xs
  | otherwise = Right $ streamFromListUtil xs

streamFromListUtil :: [a] -> Stream a
streamFromListUtil xs =
  case cycle xs of
    (v:xs') -> Cons v $ streamFromListUtil xs'

λ> streamFromList [1]
Right [1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1]

λ> streamFromList []
Left []
```